internal/mailer/validate.go

```diff
@@ -0,0 +1,290 @@
+	ctx, cancel := context.WithTimeout(ctx, validateEmailTimeout)
+	defer cancel()
+
+	// Easier control flow here to always use errgroup, it has very little
+	// overhad in comparison to the network calls it makes. The reason
+	// we run both checks concurrently is to tighten the timeout without
+	// potentially missing a call to the validation service due to a
+	// dns timeout or something more nefarious like a honeypot dns entry.
+	g := new(errgroup.Group)
+
+	// Validate the static rules first to prevent round trips on bad emails
+	// and to parse the host ahead of time.
+	if !ev.isExtendedDisabled() {
+
+		// First validate static checks such as format, known invalid hosts
+		// and any other network free checks. Running this check before we
+		// call the service will help reduce the number of calls with known
+		// invalid emails.
+		host, err := ev.validateStatic(email)
+		if err != nil {
+			return err
+		}
+
+		// Start the goroutine to validate the host.
+		g.Go(func() error { return ev.validateHost(ctx, host) })
+	}
+
+	// If the service check is not disabled we start a goroutine to run
+	// that check as well.
+	if !ev.isServiceDisabled() {
+		g.Go(func() error { return ev.validateService(ctx, email) })
+	}
+	return g.Wait()
+}
+
+// validateStatic will validate the format and do the static checks before
+// returning the host portion of the email.
+func (ev *EmailValidator) validateStatic(email string) (string, error) {
+	if !ev.extended {
+		return "", nil
+	}
+
+	ea, err := mail.ParseAddress(email)
+	if err != nil {
+		return "", ErrInvalidEmailFormat
+	}
+
+	i := strings.LastIndex(ea.Address, "@")
+	if i == -1 {
+		return "", ErrInvalidEmailFormat
+	}
+
+	// few static lookups that are typed constantly and known to be invalid.
+	if invalidEmailMap[email] {
+		return "", ErrInvalidEmailAddress
+	}
+
+	host := email[i+1:]
+	if invalidHostMap[host] {
+		return "", ErrInvalidEmailDNS
+	}
+
+	for i := range invalidHostSuffixes {
+		if strings.HasSuffix(host, invalidHostSuffixes[i]) {
+			return "", ErrInvalidEmailDNS
+		}
+	}
+
+	name := email[:i]
+	if err := ev.validateProviders(name, host); err != nil {
+		return "", err
+	}
+	return host, nil
+}
+
+func (ev *EmailValidator) validateService(ctx context.Context, email string) error {
+	if ev.isServiceDisabled() {
+		return nil
+	}
+
+	reqObject := struct {
+		EmailAddress string `json:"email"`
+	}{email}
+
+	reqData, err := json.Marshal(&reqObject)
+	if err != nil {
+		return nil
+	}
+
+	rdr := bytes.NewReader(reqData)
+	req, err := http.NewRequestWithContext(ctx, "GET", ev.serviceURL, rdr)
+	if err != nil {
+		return nil
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("apikey", ev.serviceKey)
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil
+	}
+	defer res.Body.Close()
+
+	resObject := struct {
+		Valid *bool `json:"valid"`
+	}{}
+	dec := json.NewDecoder(io.LimitReader(res.Body, 1<<5))
+	if err := dec.Decode(&resObject); err != nil {
```

Comment:
maybe we should check for the response status code before decoding the response?